features/storage/kvstore/tdbstore/TDBStore.cpp

```diff
@@ -0,0 +1,1462 @@
+{
+    uint32_t crc;
+    MbedCRC<POLY_32BIT_ANSI, 32> ct(init_crc, 0x0, true, false);
+    ct.compute(const_cast<void *>(data_buf), data_size, &crc);
+    return crc;
+}
+
+// Class member functions
+
+TDBStore::TDBStore(BlockDevice *bd) : _ram_table(0), _max_keys(0),
+    _num_keys(0), _bd(bd), _buff_bd(0),  _free_space_offset(0), _master_record_offset(0),
+    _master_record_size(0), _is_initialized(false), _active_area(0), _active_area_version(0), _size(0),
+    _prog_size(0), _work_buf(0), _key_buf(0), _variant_bd_erase_unit_size(false), _inc_set_handle(0)
+{
+}
+
+TDBStore::~TDBStore()
+{
+    deinit();
+}
+
+int TDBStore::read_area(uint8_t area, uint32_t offset, uint32_t size, void *buf)
+{
+    int os_ret = _buff_bd->read(buf, _area_params[area].address + offset, size);
+
+    if (os_ret) {
+        return MBED_ERROR_READ_FAILED;
+    }
+
+    return MBED_SUCCESS;
+}
+
+int TDBStore::write_area(uint8_t area, uint32_t offset, uint32_t size, const void *buf)
+{
+    int os_ret = _buff_bd->program(buf, _area_params[area].address + offset, size);
+    if (os_ret) {
+        return MBED_ERROR_WRITE_FAILED;
+    }
+
+    return MBED_SUCCESS;
+}
+
+int TDBStore::erase_erase_unit(uint8_t area, uint32_t offset)
+{
+    uint32_t bd_offset = _area_params[area].address + offset;
+    uint32_t eu_size = _buff_bd->get_erase_size(bd_offset);
+
+    int os_ret = _buff_bd->erase(bd_offset, eu_size);
+    if (os_ret) {
+        return MBED_ERROR_WRITE_FAILED;
+    }
+    return MBED_SUCCESS;
+}
+
+void TDBStore::calc_area_params()
+{
+    // TDBStore can't exceed 32 bits
+    bd_size_t bd_size = std::min(_bd->size(), (bd_size_t) 0x80000000L);
+
+    memset(_area_params, 0, sizeof(_area_params));
+    size_t area_0_size = 0;
+    bd_size_t prev_erase_unit_size = _bd->get_erase_size(area_0_size);
+    _variant_bd_erase_unit_size = 0;
+
+    while (area_0_size < bd_size / 2) {
+        bd_size_t erase_unit_size = _bd->get_erase_size(area_0_size);
+        _variant_bd_erase_unit_size |= (erase_unit_size != prev_erase_unit_size);
+        area_0_size += erase_unit_size;
+    }
+
+    _area_params[0].address = 0;
+    _area_params[0].size = area_0_size;
+    _area_params[1].address = area_0_size;
+    _area_params[1].size = bd_size - area_0_size;
+}
+
+
+// This function, reading a record from the BD, is used for multiple purposes:
+// - Init (scan all records, no need to return file name and data)
+// - Get (return file data)
+// - Get first/next file (check whether name matches, return name if so)
+int TDBStore::read_record(uint8_t area, uint32_t offset, char *key,
+                          void *data_buf, uint32_t data_buf_size,
+                          uint32_t &actual_data_size, size_t data_offset, bool copy_key,
+                          bool copy_data, bool check_expected_key, bool calc_hash,
+                          uint32_t &hash, uint32_t &flags, uint32_t &next_offset)
```

Comment:
Note, this is hard to read (too many parameters) - the complexity is high for this method.

Reply:
Yes, I'm aware of it. This function has many uses, with slight variations depending on the parameters, hence their vast number. Will try to think of an implementation alternative (not for 5.11 though).